tracing-appender/src/rolling.rs

```diff
@@ -97,16 +101,21 @@ pub struct RollingFileAppender {
 /// [writer]: std::io::Write
 /// [`MakeWriter`]: tracing_subscriber::fmt::writer::MakeWriter
 #[derive(Debug)]
-pub struct RollingWriter<'a>(RwLockReadGuard<'a, File>);
+pub struct RollingWriter<'a> {
+    inner: RwLockReadGuard<'a, File>,
+    current_size: &'a AtomicU64,
+}
 
 #[derive(Debug)]
 struct Inner {
     log_directory: PathBuf,
     log_filename_prefix: Option<String>,
     log_filename_suffix: Option<String>,
+    log_filename_index: Option<AtomicU64>,
```

Comment:
It's not entirely clear to me as to why `log_filename_index` and `current_size` are atomics nor am I entirely sure if they are correct. I would _naively_ prefer to wrap these two fields in a single struct that is, in turn, wrapped in a `RwLock`, but I'll let Eliza comment on this.

In any case, an explanation as to why atomics are used for this would be *really* helpful.

Reply:
This is fair, wrapping these in a `RwLock` is definitely less error-prone.

Here I used atomics for the same reason `next_date` is atomic: `RollingFileAppender` is implementing [`MakeWriter`](https://docs.rs/tracing-subscriber/0.3.17/tracing_subscriber/fmt/writer/trait.MakeWriter.html), and we need to borrow fields using a shared reference to the `RollingFileAppender`. Ideally we don’t want to pay too high synchronization costs, and there is a `RwLock<File>` that is already relied upon when swapping the file.

`current_size` is used whenever data is written to the file in the `Write` implementation of either `RollingFileAppender` or `RollingWriter<'a>`, and `fetch_add` operations are performed, so it’s basically a counter.
```rust
    fn write(&mut self, buf: &[u8]) -> io::Result<usize> {
        let written = (&*self.inner).write(buf)?;
        self.current_size.fetch_add(written, Ordering::SeqCst);
        Ok(written)
    }
```

`current_size` is reset to 0 when `refresh_writer` is called:
```rust
    fn refresh_writer(&self, now: OffsetDateTime, file: &mut File) {
        let filename = self.join_date(&now);

        if let Some(max_files) = self.max_files {
            self.prune_old_logs(max_files);
        }

        match create_writer(&self.log_directory, &filename) {
            Ok(new_file) => {
                if let Err(err) = file.flush() {
                    eprintln!("Couldn't flush previous writer: {}", err);
                }
                *file = new_file;
                self.current_size.store(0, Ordering::Release);
            }
            Err(err) => eprintln!("Couldn't create writer for logs: {}", err),
        }
    }
```
This function requires an exclusive reference to the file, and we can either obtain this exclusive reference by calling `write` on `&RwLock<File>`, or `get_mut` or `&mut RwLock<File>`. We are guaranteed to not be writing in the file at the same time because of that.

`log_filename_index` is pretty similar to `next_date`, and is actually updated only when `next_date` was updated by the current thread, in `advance_date_and_index` (previously named `advance_date`):
```rust
    fn advance_date_and_index(&self, now: OffsetDateTime, current: usize) -> bool {
        let next_date = self
            .rotation
            .next_date(&now)
            .map(|date| date.unix_timestamp() as usize)
            .unwrap_or(0);
        let next_date_updated = self
            .next_date
            .compare_exchange(current, next_date, Ordering::AcqRel, Ordering::Acquire)
            .is_ok();

        if next_date_updated {
            if let Some(index) = &self.log_filename_index {
                if current == next_date {
                    index.fetch_add(1, Ordering::SeqCst);
                } else {
                    index.store(0, Ordering::Release);
                }
            }

            true
        } else {
            false
        }
    }
```

All of this is put together in the `make_writer` implementation:
```rust
    fn make_writer(&'a self) -> Self::Writer {
        let now = self.now();

        // Should we try to roll over the log file?
        if let Some(current_time) = self.state.should_rollover(now) {
            // Did we get the right to lock the file? If not, another thread
            // did it and we can just make a writer.
            if self.state.advance_date_and_index(now, current_time) {
                self.state.refresh_writer(now, &mut self.writer.write());
            }
        }

        RollingWriter {
            inner: self.writer.read(),
            current_size: &self.state.current_size,
        }
    }
```
`advance_date_and_index` will return `true` if this thread was able to swap `next_date` with the new date, and `false` if another thread did it faster. The thread which was able to swap the value will take the exclusive write access using `RwLock::write`, and the other threads will wait for the file to be actually swapped at the `RwLock::read` call. If a thread is able to call `RwLock::read` faster, then it’s okay, we’ll simply write a few more lines in the old file before the rotation can happen.

In the `io::Write` implementation for `RollingFileAppender`, things are even more straightforward because we have an exclusive reference, and it must be the case that no other thread is accessing the atomics or the RwLock:
```rust
    fn write(&mut self, buf: &[u8]) -> io::Result<usize> {
        let now = self.now();
        let writer = self.writer.get_mut();
        if let Some(current_time) = self.state.should_rollover(now) {
            let did_cas = self.state.advance_date_and_index(now, current_time);
            debug_assert!(did_cas, "if we have &mut access to the appender, no other thread can have advanced the timestamp...");
            self.state.refresh_writer(now, writer);
        }

        let written = writer.write(buf)?;

        self.state.current_size.fetch_add(
            u64::try_from(written).expect("usize to u64 conversion"),
            Ordering::SeqCst,
        );

        Ok(written)
    }
```

If it wasn’t for the `MakeWriter`, no atomic and no RwLock would be required at all.

I hope it clarifies things. Do you want me to document that in the code directly?